### ISSUE TITLE
list_del corruption with latest Centos 7 kernel and 0.7.11-1

### ISSUE THREAD

**Issue by DrDaveD** [zfsonlinux/zfs#7933]
### System information

Type | Version/Name
 --- | --- 
Distribution Name	|  CentOS
Distribution Version	| 7.5.1804
Linux Kernel	| 3.10.0-862.11.6.el7.x86_64
Architecture	| x86_64
ZFS Version	| 0.7.11-1.el7_4
SPL Version	| 0.7.11-1.el7_4

### Describe the problem you're observing

Received a crash within the first half hour after first boot of these versions.  The first line of vmcore-dmesg.txt was
```
list_del corruption, ffff96f60e1adf00->next is LIST_POISON1 (dead000000000100)
```

### Describe how to reproduce the problem

Don't know.  We did the same commands starting up our application after reboot and it hasn't happened yet again since in the next half hour.  Will follow up if we see it again.

### Include any warning/errors/backtraces from the system logs

EDIT: added the following info to make this searchable from the GitHub web interface:

```
[  890.992193] WARNING: CPU: 2 PID: 1421 at lib/list_debug.c:53 __list_del_entry+0x63/0xd0
[  890.992194] list_del corruption, ffff96f60e1adf00->next is LIST_POISON1 (dead000000000100)
[  890.992215] Modules linked in: 8021q garp mrp stp llc nfs lockd grace sunrpc fscache tcp_diag inet_diag ip6t_REJECT nf_reject_ipv6 ip6table_filter ip6table_nat nf_conntrack_ipv6 nf_defrag_ipv6 nf_nat_ipv6 ip6_tables xt_REDIRECT nf_nat_redirect iptable_nat nf_nat_ipv4 nf_nat iptable_mangle iptable_security iptable_raw nf_conntrack_ipv4 nf_defrag_ipv4 ipt_REJECT nf_reject_ipv4 xt_comment xt_multiport xt_conntrack nf_conntrack iptable_filter zfs(POE) zunicode(POE) zavl(POE) icp(POE) iTCO_wdt iTCO_vendor_support mxm_wmi sb_edac intel_powerclamp coretemp intel_rapl iosf_mbi kvm_intel kvm irqbypass crc32_pclmul ghash_clmulni_intel aesni_intel lrw gf128mul glue_helper ablk_helper zcommon(POE) cryptd znvpair(POE) pcspkr spl(OE) ipmi_ssif ses enclosure mei_me i2c_i801 joydev sg lpc_ich mei ioatdma shpchp
[  890.992228]  ipmi_si ipmi_devintf ipmi_msghandler wmi acpi_power_meter acpi_pad ip_tables xfs libcrc32c raid1 sd_mod crc_t10dif crct10dif_generic ast i2c_algo_bit drm_kms_helper syscopyarea sysfillrect sysimgblt fb_sys_fops ttm drm ixgbe ahci mpt3sas libahci nvme mdio crct10dif_pclmul crct10dif_common ptp libata nvme_core crc32c_intel i2c_core pps_core raid_class scsi_transport_sas dca dm_mirror dm_region_hash dm_log dm_mod
[  890.992230] CPU: 2 PID: 1421 Comm: dp_sync_taskq Kdump: loaded Tainted: P        W  OE  ------------   3.10.0-862.11.6.el7.x86_64 #1
[  890.992231] Hardware name: Supermicro SSG-6048R-E1CR36L/X10DRH-iT, BIOS 3.0a 02/05/2018
[  890.992231] Call Trace:
[  890.992232]  [<ffffffffac1135d4>] dump_stack+0x19/0x1b
[  890.992235]  [<ffffffffaba945d8>] __warn+0xd8/0x100
[  890.992236]  [<ffffffffaba9465f>] warn_slowpath_fmt+0x5f/0x80
[  890.992239]  [<ffffffffabd6cf03>] __list_del_entry+0x63/0xd0
[  890.992240]  [<ffffffffabd6cf7d>] list_del+0xd/0x30
[  890.992264]  [<ffffffffc0b953d5>] multilist_sublist_remove+0x15/0x20 [zfs]
[  890.992265] systemd-journald[826]: /dev/kmsg buffer overrun, some messages lost.
[  890.992283]  [<ffffffffc0b5e094>] userquota_updates_task+0x1d4/0x580 [zfs]
[  890.992300]  [<ffffffffc0b5b2f0>] ? dmu_objset_userobjspace_upgradable+0x60/0x60 [zfs]
[  890.992316]  [<ffffffffc0b5b2f0>] ? dmu_objset_userobjspace_upgradable+0x60/0x60 [zfs]
[  890.992318]  [<ffffffffabd6cf7d>] ? list_del+0xd/0x30
[  890.992323]  [<ffffffffc04add2c>] taskq_thread+0x2ac/0x4f0 [spl]
[  890.992324] systemd-journald[826]: /dev/kmsg buffer overrun, some messages lost.
[  890.992325]  [<ffffffffabad1fe0>] ? wake_up_state+0x20/0x20
[  890.992330]  [<ffffffffc04ada80>] ? taskq_thread_spawn+0x60/0x60 [spl]
[  890.992332]  [<ffffffffababdf21>] kthread+0xd1/0xe0
[  890.992333]  [<ffffffffababde50>] ? insert_kthread_work+0x40/0x40
[  890.992336]  [<ffffffffac1255f7>] ret_from_fork_nospec_begin+0x21/0x21
[  890.992338]  [<ffffffffababde50>] ? insert_kthread_work+0x40/0x40
[  890.992339] ---[ end trace 187c6442ef979636 ]---
[  890.992340] ------------[ cut here ]------------
```

[vmcore-dmesg.txt](https://github.com/zfsonlinux/zfs/files/2402069/vmcore-dmesg.txt)

**Comment by DrDaveD** [comment]
It hasn't crashed, but a whole lot of processes got stuck doing an rm -f of a single file in the filesystem very early, so there was very little write activity.  The processes are unkillable.  I think we will have to downgrade the kernel.

**Comment by kpande** [comment]
probably a duplicate issue: https://github.com/zfsonlinux/zfs/issues/7735#issuecomment-416628385

**Comment by cvoltz** [comment]
I see a similar problem (see #7922) when running `ls -d /backup/snapshots/hourly.* | xargs --max-procs=0 --max-args=1 --replace=@ find @ -delete` to remove millions of files from nested directories (up to 35 levels deep) from a dataset which has compression enabled and dedup disabled. I was also running `rsync` on `/backup/snapshots/.sync`. After a while, `systemd-logger` was eating a ton of CPU and the system eventually became completely unresponsive. After the system was rebooted, I was able to find this repeated many times in the system log:
```
Sep 19 15:30:43 buri kernel: WARNING: CPU: 20 PID: 2092 at lib/list_debug.c:62 __list_del_entry+0x82/0xd0
Sep 19 15:30:43 buri kernel: list_del corruption. next->prev should be ffffa07f685637f0, but was dead000000000200
Sep 19 15:30:43 buri kernel: Modules linked in: ip6t_rpfilter ipt_REJECT nf_reject_ipv4 ip6t_REJECT nf_reject_ipv6 xt_conntrack ip_set nfnetlink ebtable_nat ebtable_broute bridge stp llc ip6table_nat nf_conntrack_ipv6 nf_defrag_ipv6 nf_nat_ipv6 ip6table_mangle ip6table_security ip6table_raw iptable_nat nf_conntrack_ipv4 nf_defrag_ipv4 nf_nat_ipv4 nf_nat nf_conntrack libcrc32c iptable_mangle iptable_security iptable_raw ebtable_filter ebtables ip6table_filter ip6_tables iptable_filter sb_edac intel_powerclamp coretemp intel_rapl iosf_mbi kvm_intel kvm irqbypass zfs(POE) crc32_pclmul ghash_clmulni_intel iTCO_wdt iTCO_vendor_support aesni_intel lrw gf128mul glue_helper ablk_helper cryptd zunicode(POE) zavl(POE) icp(POE) pcspkr zcommon(POE) znvpair(POE) spl(OE) lpc_ich hpilo hpwdt joydev ipmi_ssif ipmi_si sg ipmi_devintf
Sep 19 15:30:43 buri kernel: ipmi_msghandler acpi_power_meter wmi ioatdma shpchp pcc_cpufreq ip_tables ext4 mbcache jbd2 sr_mod cdrom sd_mod crc_t10dif crct10dif_generic uas usb_storage mgag200 i2c_algo_bit drm_kms_helper syscopyarea sysfillrect sysimgblt fb_sys_fops ttm drm ixgbe crct10dif_pclmul crct10dif_common tg3 crc32c_intel serio_raw hpsa mdio i2c_core dca ptp scsi_transport_sas pps_core dm_mirror dm_region_hash dm_log dm_mod
Sep 19 15:30:43 buri kernel: CPU: 20 PID: 2092 Comm: dp_sync_taskq Kdump: loaded Tainted: P           OE  ------------   3.10.0-862.11.6.el7.x86_64 #1
Sep 19 15:30:43 buri kernel: Hardware name: HP ProLiant DL380p Gen8, BIOS P70 01/22/2018
Sep 19 15:30:43 buri kernel: Call Trace:
Sep 19 15:30:43 buri kernel: [<ffffffffaa5135d4>] dump_stack+0x19/0x1b
Sep 19 15:30:43 buri kernel: [<ffffffffa9e945d8>] __warn+0xd8/0x100
Sep 19 15:30:43 buri kernel: [<ffffffffa9e9465f>] warn_slowpath_fmt+0x5f/0x80
Sep 19 15:30:43 buri kernel: [<ffffffffaa16cf22>] __list_del_entry+0x82/0xd0
Sep 19 15:30:43 buri kernel: [<ffffffffaa16cf7d>] list_del+0xd/0x30
Sep 19 15:30:43 buri kernel: [<ffffffffc09ca3d5>] multilist_sublist_remove+0x15/0x20 [zfs]
Sep 19 15:30:43 buri kernel: [<ffffffffc0993094>] userquota_updates_task+0x1d4/0x580 [zfs]
Sep 19 15:30:43 buri kernel: [<ffffffffc09902f0>] ? dmu_objset_userobjspace_upgradable+0x60/0x60 [zfs]
Sep 19 15:30:43 buri kernel: [<ffffffffc09902f0>] ? dmu_objset_userobjspace_upgradable+0x60/0x60 [zfs]
Sep 19 15:30:43 buri kernel: [<ffffffffaa16cf7d>] ? list_del+0xd/0x30
Sep 19 15:30:43 buri kernel: [<ffffffffc04ecd2c>] taskq_thread+0x2ac/0x4f0 [spl]
Sep 19 15:30:43 buri kernel: [<ffffffffa9ed1fe0>] ? wake_up_state+0x20/0x20
Sep 19 15:30:43 buri kernel: [<ffffffffc04eca80>] ? taskq_thread_spawn+0x60/0x60 [spl]
Sep 19 15:30:43 buri kernel: [<ffffffffa9ebdf21>] kthread+0xd1/0xe0
Sep 19 15:30:43 buri kernel: [<ffffffffa9ebde50>] ? insert_kthread_work+0x40/0x40
Sep 19 15:30:43 buri kernel: [<ffffffffaa5255f7>] ret_from_fork_nospec_begin+0x21/0x21
Sep 19 15:30:43 buri kernel: [<ffffffffa9ebde50>] ? insert_kthread_work+0x40/0x40
Sep 19 15:30:43 buri kernel: ---[ end trace fb2fe7f782509b51 ]---
```
Apparently, this got generated so many times that `systemd-logger` started missing kernel messages and, presumably, that was why it was taking so much CPU.

**Comment by shodanshok** [comment]
I tried to reproduce the problem with very heavy, concurrent `fs_mark` and `fio` workloads but with no avail. I am on CentOS 7.5, kernel 3.10.0-862.11.6.el7.x86_64, zfs-0.7.11-1.el7_5.x86_64 and spl-0.7.11-1.el7_5.x86_64. 

I see the OP is using ZFS and SPL version 0.7.11-1.el7_4. Any changes to update?

**Comment by DrDaveD** [comment]
Our yum.repos.d file was accidentally loading the el7_4 packages even though we are on CentOS 7.5, but I am assuming that doesn't matter because the kernel number matched.  We have now switched to the el7_5 yum repository but have backed it down to an older kernel and version 0.7.9-1.el7_5 packages and didn't try it with 0.7.11-1.el7_5.  Does anybody know if there might be a significant difference between 0.7.11-1.el7_4 and 0.7.11-1.el7_5?

**Comment by tonyhutter** [comment]
>  Does anybody know if there might be a significant difference between 0.7.11-1.el7_4 and 0.7.11-1.el7_5?

Funny you should mention this.  The Centos 7.4 0.7.11 kmod rpms were incorrectly building against the Centos 7.5 kernel, but this has since been fixed (#7956).

**Comment by cvoltz** [comment]
Note that in my case, I was running CentOS 7.5 with the dkms version of 0.7.11-1.el7_5 and the system was a clean install with 7.5 so there was no possibility for 7.4/7.5 confusion. On a dataset with dedup off, I have successfully done the same command that caused a crash before on the dataset with dedup on so I wonder if dedup being on could be a factor. Tthe datasets are similar but different so that could be a factor. @DrDaveD did you have dedup on or off on your dataset?

**Comment by DrDaveD** [comment]
Our dedup was off.

**Comment by tonyhutter** [comment]
@cvoltz https://github.com/zfsonlinux/zfs/issues/7997 looks to have the same del_list corruption stacktrace as you.  I'd like to try to reproduce the issue on our lustre test cluster.  Can you give us any more detail on the dataset you were trying to delete (dataset size, file size, etc)?

**Comment by cvoltz** [comment]
The dataset currently has 6.72 TB of de-duped data from the `/etc`, `/home`, `/opt`, `/usr/local`, `/var/lib`, `/var/log`, `/var/named`, and `/var/www` directories of several machines in 119,434,653 files in directories as deep as 35 levels. Most of the files are hard links to other files as new snapshots (not ZFS snapshots) are generated by `rsnapshot` using commands like `cp -al /backup/snapshots/.sync /backup/snapshots/hourly.0`. Here is the distribution of the file sizes (where size is in bytes and count is the number of files with that size or larger but smaller then the next size):
```
size         count
           0 1157117
           1 322651
           2 283200
           4 692243
           8 865936
          16 451844
          32 2281929
          64 2592941
         128 4350013
         256 15884891
         512 7617199
        1024 8248393
        2048 8981136
        4096 8219888
        8192 7495136
       16384 7558788
       32768 7755821
       65536 12748289
      131072 10859115
      262144 2783969
      524288 2177289
     1048576 2046111
     2097152 1722066
     4194304 1125459
     8388608 630353
    16777216 262524
    33554432 158987
    67108864 61518
   134217728 59377
   268435456 24303
   536870912 5046
  1073741824 2978
  2147483648 3766
  4294967296 3923
  8589934592 503
 17179869184 47
 34359738368 30
137438953472 34
```
As an aside, it took a little over 6 hours to stat all the files to generate the file size histogram. The systems being backed up are running ext4, and while they do have less files, they take a **lot** less time to stat. For comparison, on one system, it took 34s to run `ls -lR . > /dev/null` on 1,796,732 files in 264,139 directories. A subset of those files on the ZFS filesystem has 1,701,649 files in 254,678 directories and the same command took 195s (which is 5.7x).

I had multiple directories (`hourly.0`, `hourly.1`, ...), which each had a number of files and directories similar to the ones I benchmarked (e.g., 1,701,649 files in 254,678 directories), which I was deleting simultaneously when I got the stack trace.
